pathdb/datastax/src/main/java/org/commonjava/storage/pathmapped/pathdb/datastax/CassandraPathDB.java

```diff
@@ -138,13 +136,93 @@ public Session getSession()
+    {
+        if ( recursive )
+        {
+            List<PathMap> ret = new ArrayList<>();
+            traverse( fileSystem, path, pathMap -> ret.add( pathMap ), limit );
+            return ret;
+        }
+        else
+        {
+            String parentPath = PathMapUtils.normalizeParentPath( path );
+            Result<DtxPathMap> ret = boundAndRunListQuery( fileSystem, parentPath );
+            return new ArrayList<>( ret.all() );
+        }
+    }
+
+    private Result<DtxPathMap> boundAndRunListQuery( String fileSystem, String parentPath )
+    {
         BoundStatement bound = preparedListQuery.bind( fileSystem, parentPath );
         ResultSet result = session.execute( bound );
-        Result<DtxPathMap> ret = pathMapMapper.map( result );
-        return new ArrayList<>( ret.all() );
+        return pathMapMapper.map( result );
+    }
+
+    private final static DtxPathMap FAKE_ROOT_OBJ = new DtxPathMap(); // if path is ROOT_DIR, use a FAKE_ROOT_OBJ
+
+    // To avoid listing huge file system and OOM we throw exception when the results exceeds limit
+    private void traverse( String fileSystem, String path, Consumer<PathMap> consumer, int limit )
+    {
+        logger.debug( "Traverse fileSystem: {}, path: {}", fileSystem, path );
+
+        DtxPathMap root;
+        if ( ROOT_DIR.equals( path ) )
+        {
+            root = FAKE_ROOT_OBJ;
+        }
+        else
+        {
+            if ( !path.endsWith( "/" ) )
+            {
+                path += "/";
+            }
+            String parentPath = PathMapUtils.getParentPath( path );
+            String filename = PathMapUtils.getFilename( path );
+            root = pathMapMapper.get( fileSystem, parentPath, filename );
+            if ( root == null )
+            {
+                logger.debug( "Root not found, fileSystem: {}, parentPath: {}, filename: {}", fileSystem, parentPath, filename );
+                return;
+            }
+        }
+
+        TreeTraverser<PathMap> traverser = new TreeTraverser<PathMap>()
+        {
+            @Override
+            public Iterable<PathMap> children( PathMap cur )
+            {
+                String parentPath;
+                if ( cur == FAKE_ROOT_OBJ )
+                {
+                    parentPath = ROOT_DIR;
+                }
+                else
+                {
+                    parentPath = Paths.get( cur.getParentPath(), cur.getFilename() ).toString();
+                }
+                Result<DtxPathMap> ret = boundAndRunListQuery( fileSystem, parentPath );
+                return new ArrayList<>( ret.all() );
+            }
+        };
+        AtomicInteger count = new AtomicInteger( 0 );
+        traverser.preOrderTraversal( root ).forEach( dtxPathMap -> {
+            if ( limit > 0 && count.incrementAndGet() > limit )
+            {
+                throw new RuntimeException( "Exceeds results limit " + limit );
```

Comment:
Indy doesn't have this use case. 
In the past, when we found a file was broken (e.g. maven-metadata.xml), we deleted it from disk. With path-mapped, we are going to traverse the files under a repo or a folder, get the file, or delete it by REST api. We have a JIRA NOS-2031 for that. And I will submit an Indy pr too.
The out-of-limit problem is not a normal use case so I did not use a checked type. In other words, this tool is for our internal use. Just to prevent us unintentionally listing something like pnc-builds.

Reply:
Ok. I think this pr can be merged. But for further consideration some more traceable log in exception is still a good practice for us to follow.

Reply:
Yep. Maybe we need a checked exception in the future. We can update then.